jt_notes.md

```diff
@@ -0,0 +1,127 @@
+
+- **IMPORTANT:** the same questions apply to the Flask app. As per my comment in app.py, I am starting to suspect that this could be running concurrent threads. That needs careful thought and documenting, because it gets very complicated.
+
+- More generally, *as far as I can see*, there doesn't seem to be to be a critical need to be using [sequential] threads here at all (though it doesn't do any harm). Out of interest, is this code just adapted from some online example? I suspect it makes a lot more sense in the case where there is *more than one* client involved (but I don't think that's really an issue for us in reality).
+
+**Out of interest:** I find the code formatting [newline after "(", and ")" on a line of its own, sometimes doing this in rather superfluous contexts] really weird. Is this an established coding style, or just what Alex does!? It's certainly one way to keep to line-width limits, but not the one I would choose - and line limits are violated elsewhere! Extreme example:
+
+                                        self.trigger_fluorescence_image_capture(
+
+                                                                                                                        current_time + timeToWaitInSecs
+
+                                                                                                                        )
+
+Hmm... although I can see that, in the absence of curly brackets for statement blocks, like in C, it helps with readability (in terms of the indenting layout) in cases like this:
+
+if (period != -1
+
+and len(periods) >= (5 + (2 * settings["numExtraRefFrames"]))
+
+and period > 6
+
+and (len(periods) - 1 - settings["numExtraRefFrames"]) > 0
+
+and (periods[len(periods) - 1 - settings["numExtraRefFrames"]]) > 6
+
+                ):
+
+                        periodToUse = periods[len(periods) - 1 - settings["numExtraRefFrames"]]
+
+------------------------------------ **Architecture sketch ------------------------------------**
+
+I am still working on this, but I think there are benefits to having things in their own (concurrent) threads, which in python effectively means they are separate *processes* that must pass messages to each other (or be spawned and return asynchronously). For interfacing with M^2 system, I want a Sync Analyzer process that is driven by [low-level loopback IP?] messages sent back and forth between it and a client. On the RPi, we would also have a Sync Controller process, which would probably be the existing Flask loop, to take care of interfacing with camera/pins hardware. This Sync Controller would be in asynchronous communication with Sync Analyzer, but could also potentially be receiving GUI input from a web client interface.
```

Comment:
So, I see a similar thing using GET/POST methods. I.e. clients (Controllers), like the M^2 hardware, will just GET current state of the Analyser and then will control any action 'locally', e.g. take image at time T, and the GUI will POST, for example, settings or user interrupts. I guess the question here is 'is just listening for GETs good enough', i.e. do we need a more active push from the Analyser rather than just a pull from the Controller?

Reply:
I think for the most part it could be handled using client-driven messages. For example, predictions will be generated _in response to_ a new frame being supplied to the analyzer. We should think carefully, but offhand I can't think of anything that would actually require a push. The LTU makes things slightly complicated though, because of the time delay involved in computing the LTU. A properly-responsive live system would definitely need a Sync Controller client that was able to cope with async responses, i.e. it sends an LTU message but does not block waiting for the response to come back (and does other things in the meantime, including sending more frames to the Analyzer).

Reply:
As a reality check for myself: there is nothing that the ObjC Sync Analyzer does that is not in some sense client-driven. It posts back a diverse range of different "events" (effectively, response messages), but ultimately any code that it executes is run because a new frame has become available. So in conclusion: the response format might get complicated, but it should work.